tests/src/sanity.rs

```diff
@@ -0,0 +1,326 @@
+    #[tokio::test]
+    // starts the node(express.js) api server, sends four different requests, validates data,
+    // stops the server and validates if the agent job and pod are deleted
+    async fn test_complete_node_api() {
+        let client = setup_kube_client().await;
+
+        let service_url = get_service_url(&client, "default").await.unwrap();
+        let pod_name = get_nginx_pod_name(&client, "default").await.unwrap();
+        let command = vec!["node", "node-e2e/app.js"];
+        let env: HashMap<&str, &str> = HashMap::new(); // for adding more environment variables
+        let mut server = start_node_server(&pod_name, command, env);
+
+        let mut stderr_reader = BufReader::new(server.stderr.take().unwrap());
+        let child_stdout = server.stdout.take().unwrap();
+
+        setup_panic_hook();
+
+        tokio::spawn(async move {
+            loop {
+                let mut error_stream = String::new();
+                // the following is a blocking call, so we need to spawn a task to read the stderr
+                stderr_reader.read_line(&mut error_stream).await.unwrap();
+                // Todo(investigate): it looks like the task reaches the panic when not kept in a
+                // loop i.e. the buffer reads an empty string:  `thread 'main'
+                // panicked at 'Error: '`
+                if !error_stream.is_empty() {
+                    panic!("Error: {}", error_stream);
+                }
+            }
+        });
+
+        // since we are reading from the stdout, we could block at any point in case the server
+        // does not write to its stdout, so we need a timeout here
+        let validation_duration = Duration::from_secs(30);
+        timeout(
+            validation_duration,
+            validate_requests(child_stdout, service_url.as_str(), &mut server),
+        )
+        .await
+        .unwrap();
+
+        // look for the agent job and pod
+        let jobs_api: Api<Job> = Api::namespaced(client.clone(), "default");
+        let jobs = jobs_api.list(&ListParams::default()).await.unwrap();
+        // assuming only one job is running
+        // to make the tests parallel we need to figure a way to get the exact job name when len() >
+        // 1
+        assert_eq!(jobs.items.len(), 1);
+        let job_name = jobs.items[0].metadata.name.clone().unwrap();
+
+        let pods_api: Api<Pod> = Api::namespaced(client.clone(), "default");
+        let pods = pods_api.list(&ListParams::default()).await.unwrap();
+        let pod_name = pods
+            .items
+            .iter()
+            .filter_map(|pod| {
+                if pod.metadata.name.clone().unwrap().contains(&job_name) {
+                    Some(pod.metadata.name.clone().unwrap())
+                } else {
+                    None
+                }
+            })
+            .next()
+            .unwrap();
+        assert!(pod_name.contains(&job_name));
+        sleep(Duration::from_secs(35)).await;
```

Comment:
As discussed (I think):
We want to try and not sleep the whole 35 seconds, if possible. So lets do something like this:

counter = 0
job_len = 1
while counter =< 35 && job_len > 0
{
job_len = pods_api...
sleep(1)
counter++
}
assert(job_len, 0)